packages/aws-cdk/lib/toolkit/error.ts

```diff
@@ -0,0 +1,48 @@
+const TOOLKIT_ERROR_SYMBOL = Symbol.for('@aws-cdk/core.TooklitError');
+const AUTHENTICATION_ERROR_SYMBOL = Symbol.for('@aws-cdk/core.AuthenticationError');
+
+/**
+ * Represents a general toolkit error in the AWS CDK Toolkit.
+ */
+class ToolkitError extends Error {
+  /**
+   * Determines if a given error is an instance of ToolkitError.
+   */
+  public static isToolkitError(x: any): x is ToolkitError {
+    return x !== null && typeof(x) === 'object' && TOOLKIT_ERROR_SYMBOL in x;
+  }
+
+  /**
+   * Determines if a given error is an instance of AuthenticationError.
+   */
+  public static isAuthenticationError(x: any): x is AuthenticationError {
+    return this.isToolkitError(x) && AUTHENTICATION_ERROR_SYMBOL in x;
+  }
+
+  /**
+   * The type of the error, defaults to "toolkit".
+   */
+  public readonly type: string;
+
+  constructor(message: string, type: string = 'toolkit') {
+    super(message);
+    Object.setPrototypeOf(this, ToolkitError.prototype);
+    Object.defineProperty(this, TOOLKIT_ERROR_SYMBOL, { value: true });
+    this.name = new.target.name;
```

Comment:
[just for my benefit]: what is this? i've not seen `new.target.name` before.

Reply:
I'm following the implementation from `ConstructError`: https://github.com/aws/aws-cdk/blob/8718767c5bf9fce33ea80e6c7d3bb136c79dbf7d/packages/aws-cdk-lib/core/lib/errors.ts#L77

Apparently it is to dynamically set the `name` property to the correct class name, even when it's subclassed: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new.target